src/storage/hummock_trace/src/collector.rs

```diff
@@ -0,0 +1,506 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env;
+use std::fs::{create_dir_all, OpenOptions};
+use std::io::BufWriter;
+use std::ops::Bound;
+use std::path::Path;
+use std::sync::atomic::AtomicU64;
+use std::sync::LazyLock;
+
+use bincode::{Decode, Encode};
+use bytes::Bytes;
+use parking_lot::Mutex;
+use risingwave_hummock_sdk::HummockReadEpoch;
+use risingwave_pb::meta::SubscribeResponse;
+use tokio::sync::mpsc::{
+    unbounded_channel as channel, UnboundedReceiver as Receiver, UnboundedSender as Sender,
+};
+use tokio::task_local;
+
+use crate::write::{TraceWriter, TraceWriterImpl};
+use crate::{
+    ConcurrentIdGenerator, Operation, OperationResult, Record, RecordId, RecordIdGenerator,
+    TracedNewLocalOptions, TracedReadOptions, TracedSubResp, UniqueIdGenerator,
+};
+
+// Global collector instance used for trace collection
+static GLOBAL_COLLECTOR: LazyLock<GlobalCollector> = LazyLock::new(GlobalCollector::new);
+
+// Global record ID generator for generating unique record IDs
+static GLOBAL_RECORD_ID: LazyLock<RecordIdGenerator> =
+    LazyLock::new(|| UniqueIdGenerator::new(AtomicU64::new(0)));
+
+// Flag indicating whether trace should be used
+static SHOULD_USE_TRACE: LazyLock<bool> = LazyLock::new(set_should_use_trace);
+
+// Concurrent record ID generator for generating unique record IDs in concurrent environments
+pub static CONCURRENT_ID: LazyLock<ConcurrentIdGenerator> =
+    LazyLock::new(|| UniqueIdGenerator::new(AtomicU64::new(0)));
+
+pub const USE_TRACE: &str = "USE_HM_TRACE"; // Environment variable name for enabling trace
+const LOG_PATH: &str = "HM_TRACE_PATH"; // Environment variable name for specifying trace log path
+const DEFAULT_PATH: &str = ".trace/hummock.ht"; // Default trace log path
+const WRITER_BUFFER_SIZE: usize = 1024; // Buffer size for trace writer
+
+/// Returns whether trace should be used based on the environment variable
+pub fn should_use_trace() -> bool {
+    *SHOULD_USE_TRACE
+}
+
+/// Sets the value of the `SHOULD_USE_TRACE` flag based on the `USE_TRACE` environment variable
+fn set_should_use_trace() -> bool {
+    match std::env::var(USE_TRACE) {
+        Ok(v) => v.parse().unwrap_or(false),
+        Err(_) => false,
+    }
+}
+
+/// Initialize the `GLOBAL_COLLECTOR` with configured log file
+pub fn init_collector() {
+    tokio::spawn(async move {
```

Comment:
We'd better use a separate tokio runtime to reduce resource contention with the normal runtime.